nixos/modules/services/databases/postgresql.nix

```diff
@@ -350,17 +315,8 @@ in
               ''}
               rm -f "${cfg.dataDir}/.first_startup"
             fi
-          '' + optionalString (cfg.ensureDatabases != []) ''
-            ${concatMapStrings (database: ''
-              $PSQL -tAc "SELECT 1 FROM pg_database WHERE datname = '${database}'" | grep -q 1 || $PSQL -tAc 'CREATE DATABASE "${database}"'
-            '') cfg.ensureDatabases}
-          '' + ''
-            ${concatMapStrings (user: ''
-              $PSQL -tAc "SELECT 1 FROM pg_roles WHERE rolname='${user.name}'" | grep -q 1 || $PSQL -tAc 'CREATE USER "${user.name}"'
-              ${concatStringsSep "\n" (mapAttrsToList (database: permission: ''
-                $PSQL -tAc 'GRANT ${permission} ON ${database} TO "${user.name}"'
-              '') user.ensurePermissions)}
-            '') cfg.ensureUsers}
+
+            $PSQL -f ${statementsSql}
```

Comment:
Would it make sense to wrap this in some kind of sql transaction to make sure that if something fails, the entire change is rolled back?

Reply:
Maybe. It gives less flexibility if we do that. I assume the desired outcome is more reliability? We don't want to advertise this whole thing as reliable, so maybe leave wrapping in a transaction to people who use the `statements` option. Since everything should be idempotent in theory transaction shouldn't be required...